infrastructures/infrastructure-azure/src/main/java/org/ow2/proactive/resourcemanager/nodesource/billing/AzureBillingResourceUsage.java

```diff
@@ -196,69 +204,156 @@ String getLastResourceUsageHistory(AzureBillingCredentials azureBillingCredentia
+                             currentResourceUri + " matches " + this.resourceUriRegex + " ? " +
+                             currentResourceUri.matches(this.resourceUriRegex));
+
+                // Here, we have a resource usage per hour (i.e.  (startDateTime) [8:00,9:00], [9:00,10:00], [10:00,11:00] (endDateTime))
+                // In case of multiple VM deployed, VM names = <node source name><number>
+                if (currentResourceUri.matches(this.resourceUriRegex)) {
+
+                    LOGGER.debug("AzureBillingResourceUsage updateResourceUsageInfosOrGetMetersIds (update) considering resource " +
+                                 resourceProperties);
+
+                    double resourceQuantityInThatHour = resourceProperties.get("quantity").getAsDouble();
+                    String meterId = resourceProperties.get("meterId").getAsString();
+
+                    // Store meterId to make AzureBillingRateCard store meter rates with ids in meterIdsSet
+                    this.metersIds.add(meterId);
+
+                    // Get the meter rates of meterId
+                    LinkedHashMap<String, Double> meterRates = metersRates.get(meterId);
+
+                    if (meterRates == null) {
+                        // It should never happens but in that case do not consider this resource consumption in that period for the global cost
+                        LOGGER.debug("AzureBillingResourceUsage updateResourceUsageInfosOrGetMetersIds (update) cannot retrieve meter rate for " +
+                                     meterId + ". The global usage cost will not include the resource " +
+                                     currentResourceUri + " at this period.");
+                        continue;
+                    } else {
+                        LOGGER.debug("AzureBillingResourceUsage updateResourceUsageInfosOrGetMetersIds (update) retrieved meter rate for meterId " +
+                                     meterId + " meterRates " + meterRates);
+                    }
+
+                    double costInThatHour = computeCostInThatHour(resourceQuantityInThatHour, meterRates);
+                    this.globalCost += costInThatHour;
+                    this.budgetPercentage = this.globalCost * 100 / this.budget;
+
+                    LOGGER.debug("AzureBillingResourceUsage updateResourceUsageInfosOrGetMetersIds (update) (in while) currentResourceUri " +
+                                 currentResourceUri + " costInThatHour " + costInThatHour + " (now this.globalCost=" +
+                                 this.globalCost + ") for [" + resourceProperties.get("usageStartTime").getAsString() +
+                                 ";" + resourceProperties.get("usageEndTime").getAsString() + "]");
+
+                } // if (currentResourceUri.matches(this.resourceUriRegex))
+            } // while (resourceUsageIterator.hasNext())
```

Comment:
Remove commented code or explain why you keep it

Reply:
These are comments .. will add END OF before them in the next commit.